refactor: tagging related cleanup

benesjan opened the pull request:
I started working on improving unconstrained tagging as proposed in [this forum post](https://forum.aztec.network/t/on-note-discovery-and-index-coordination/7165) and the initial step in implementing that is the following:

1. store the tagging indexes requested during execution in a cache (just like we do with notes),
2. when the execution is finished take the cache and along with a tx hash dump it into db.

Given that the cache will reside in `PrivateExecutionOracle` I now need to orchestrate the sync there.

When trying to do that (to be done in a [PR up the stack](https://github.com/AztecProtocol/aztec-packages/pull/17445)) I realized that the current tagging-related API was cumbersome and this made me do the following in this PR:
1. Have the API accept directionalAppTaggingSecret on the input instead of appTaggingSecret and recipient as now I can just compute it in the `PrivateExecutionOracle` and request info based on that,
2. I moved the tagging functionality from stdlib to PXE as there was no reason to have it in stdlib,
3. I simplified a bunch of random things.

This PR is a bit random so if something is not clear LMK.

benesjan commented:
* **#19290** <a href="https://app.graphite.com/github/pr/AztecProtocol/aztec-packages/19290?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#19289** <a href="https://app.graphite.com/github/pr/AztecProtocol/aztec-packages/19289?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#19233** <a href="https://app.graphite.com/github/pr/AztecProtocol/aztec-packages/19233?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#19125** <a href="https://app.graphite.com/github/pr/AztecProtocol/aztec-packages/19125?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#19161** <a href="https://app.graphite.com/github/pr/AztecProtocol/aztec-packages/19161?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#19158** <a href="https://app.graphite.com/github/pr/AztecProtocol/aztec-packages/19158?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#19157** <a href="https://app.graphite.com/github/pr/AztecProtocol/aztec-packages/19157?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#19030** <a href="https://app.graphite.com/github/pr/AztecProtocol/aztec-packages/19030?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#17611** <a href="https://app.graphite.com/github/pr/AztecProtocol/aztec-packages/17611?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#17568** <a href="https://app.graphite.com/github/pr/AztecProtocol/aztec-packages/17568?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#17534** <a href="https://app.graphite.com/github/pr/AztecProtocol/aztec-packages/17534?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#17445** <a href="https://app.graphite.com/github/pr/AztecProtocol/aztec-packages/17445?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#17451** <a href="https://app.graphite.com/github/pr/AztecProtocol/aztec-packages/17451?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 👈 <a href="https://app.graphite.com/github/pr/AztecProtocol/aztec-packages/17451?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* `next`


<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.
<!-- Current dependencies on/for this PR: -->

benesjan commented on yarn-project/pxe/src/contract_function_simulator/oracle/private_execution_oracle.ts:
```diff
@@ -192,7 +192,17 @@ export class PrivateExecutionOracle extends UtilityExecutionOracle implements IP
    * @returns An app tag to be used in a log.
    */
   public async privateGetNextAppTagAsSender(sender: AztecAddress, recipient: AztecAddress): Promise<Fr> {
```
This function will now be doing the orchestration because we will be merging here the info from both ExecutionTaggingIndexCache and the db.

benesjan commented on yarn-project/stdlib/src/logs/indexed_tagging_secret.ts:
```diff
@@ -0,0 +1,6 @@
+import type { Fr } from '@aztec/foundation/fields';
+
+export type IndexedTaggingSecret = {
```
This type now became only a simple container and it no longer implements any functionality. This made sense as it was too common that we just constructed the original type and then immediately just computed the tag from it. Making it cubersome to have the functionality associated on the type.

benesjan commented on yarn-project/stdlib/src/keys/derivation.ts:
```diff
@@ -121,27 +120,3 @@ export async function deriveKeys(secretKey: Fr) {
   };
 }
-
-// Returns shared tagging secret computed with Diffie-Hellman key exchange.
-async function computeTaggingSecretPoint(knownAddress: CompleteAddress, ivsk: Fq, externalAddress: AztecAddress) {
-  const knownPreaddress = await computePreaddress(await knownAddress.publicKeys.hash(), knownAddress.partialAddress);
-  // TODO: #8970 - Computation of address point from x coordinate might fail
-  const externalAddressPoint = await externalAddress.toAddressPoint();
-  const curve = new Grumpkin();
-  // Given A (known complete address) -> B (external address) and h == preaddress
-  // Compute shared secret as S = (h_A + ivsk_A) * Addr_Point_B
-
-  // Beware! h_a + ivsk_a (also known as the address secret) can lead to an address point with a negative y-coordinate, since there's two possible candidates
-  // computeAddressSecret takes care of selecting the one that leads to a positive y-coordinate, which is the only valid address point
-  return curve.mul(externalAddressPoint, await computeAddressSecret(knownPreaddress, ivsk));
-}
-
-export async function computeAppTaggingSecret(
-  knownAddress: CompleteAddress,
-  ivsk: Fq,
-  externalAddress: AztecAddress,
-  app: AztecAddress,
-) {
-  const taggingSecretPoint = await computeTaggingSecretPoint(knownAddress, ivsk, externalAddress);
-  return poseidon2Hash([taggingSecretPoint.x, taggingSecretPoint.y, app]);
-}
```
This has been moved to `yarn-project/pxe/src/contract_function_simulator/tagging_utils.ts`

benesjan commented on yarn-project/pxe/src/contract_function_simulator/execution_data_provider.ts:
```diff
@@ -214,13 +214,36 @@ export interface ExecutionDataProvider {
    * @param recipient - The address receiving the note
+   * @returns The directional app tagging secret
+   */
+  calculateDirectionalAppTaggingSecret(
```
I needed to expose this because now the `privateGetNextAppTagAsSender` needs it.

benesjan commented on yarn-project/pxe/src/contract_function_simulator/execution_data_provider.ts:
```diff
@@ -214,13 +214,36 @@ export interface ExecutionDataProvider {
+   * because the function performs second round of siloing which is necessary because kernels do it as well (they silo
+   * first field of the private log which corresponds to the tag).
+   */
+  syncTaggedLogsAsSender(directionalAppTaggingSecret: Fr, contractAddress: AztecAddress): Promise<void>;
```
I needed to expose this because now the privateGetNextAppTagAsSender calls this.

nventuro replied:
We should probably make this just be a standalone fn in the tagging module (along with IndexedTaggingSecret) instead of adding yet another method to this just to have easy access to the stores and node. It'd also make the exact dependencies more obvious, and make testing _much_ easier.

benesjan replied:
Agree. Would do this in a PR up the stack if you are fine with it.

benesjan commented on yarn-project/pxe/src/contract_function_simulator/pxe_oracle_interface.ts:
```diff
@@ -327,30 +329,30 @@ export class PXEOracleInterface implements ExecutionDataProvider {
+    }));
   }
 
-  /**
-   * Updates the local index of the shared tagging secret of a sender / recipient pair
-   * if a log with a larger index is found from the node.
-   * @param contractAddress - The address of the contract that the logs are tagged for
-   * @param sender - The address of the sender, we must know the sender's ivsk_m.
-   * @param recipient - The address of the recipient.
-   * TODO: This is used only withing PXEOracleInterface and tests so we most likely just want to hide this.
-   */
```
The docs are in the interface now so I just dropped it from here. We will most likely drop that interface soon so doesn't really matter.

benesjan commented on yarn-project/pxe/src/contract_function_simulator/pxe_oracle_interface.ts:
```diff
@@ -388,20 +389,20 @@ export class PXEOracleInterface implements ExecutionDataProvider {
+      await this.taggingDataProvider.setTaggingSecretIndexAsSender(directionalAppTaggingSecret, currentIndex);
 
-      this.log.debug(`Syncing logs for sender ${sender} at contract ${contractName}(${contractAddress})`, {
-        sender,
-        secret: appTaggingSecret,
-        index: currentIndex,
-        contractName,
-        contractAddress,
-      });
+      this.log.debug(
+        `Syncing logs for secret ${directionalAppTaggingSecret.toString()} at contract ${contractName}(${contractAddress})`,
+        {
+          index: currentIndex,
+          contractName,
+          contractAddress,
+        },
+      );
     } else {
-      this.log.debug(`No new logs found for sender ${sender} at contract ${contractName}(${contractAddress})`);
+      this.log.debug(
+        `No new logs found for secret ${directionalAppTaggingSecret.toString()} at contract ${contractName}(${contractAddress})`,
+      );
```
These logs now became worse but I print out enough info in PrivateExecutionOracle that still makes it possible to associate the directional tagging secret with sender.

benesjan commented on yarn-project/pxe/src/contract_function_simulator/tagging_utils.ts:
```diff
@@ -1,32 +1,71 @@
+  // Beware! h_a + ivsk_a (also known as the address secret) can lead to an address point with a negative y-coordinate, since there's two possible candidates
+  // computeAddressSecret takes care of selecting the one that leads to a positive y-coordinate, which is the only valid address point
+  return curve.mul(externalAddressPoint, await computeAddressSecret(knownPreaddress, ivsk));
+}
```
This has been moved here from stdlib untouched.

benesjan commented on yarn-project/pxe/src/contract_function_simulator/tagging_utils.ts:
```diff
@@ -1,32 +1,71 @@
+  const taggingSecretPoint = await computeTaggingSecretPoint(knownAddress, ivsk, externalAddress);
+  const appTaggingSecret = await poseidon2Hash([taggingSecretPoint.x, taggingSecretPoint.y, app]);
+  return poseidon2Hash([appTaggingSecret, recipient]);
+}
```
This has been modified such that it returns directional app tagging secret and not app tagging secret. We never cared about the non-directional secret so made sense to expose function just for the directional one.

benesjan commented on yarn-project/pxe/src/contract_function_simulator/tagging_utils.ts:
```diff
@@ -1,32 +1,71 @@
+
+export async function computeSiloedTag(directionalAppTaggingSecret: Fr, index: number, app: AztecAddress) {
+  return poseidon2Hash([app, await computeTag(directionalAppTaggingSecret, index)]);
+}
```
This has been moved from what was originally IndexedTaggingSecret. Having it on the type was not really convenient because we always just constructed the IndexedTaggingSecret instance on one line and then immediately computed the tag.

nventuro replied:
I am not sure I'm a big fan of this. On the one hand, having any kind of tag code in `contract_function_simulator` seems a bit weird - there's nothing about it that's related to a simulation. On the other, it means we lose all type information and just end up with `Fr`s that need a very complicated name in order to tell us what they are, and there's high likelihood of error.

What if we had types for `TaggingSharedSecret`, the siloed shared secret, the directional siloed shared secret, and then finally the tag? Where each type has a fn that receives the next value (app addr, sender/recipient addr, index) and returns the next type. That sounds like it'd make the construction of the tag easier to follow, would make errors more obvious, would require less verbose names, etc., and it should be fairly easy. And yes, stdlib is probably not a good place for that, but contract_fn_simu is also a bad one. We can just have it in some directory in pxe if truly noone else uses the tag struct.

benesjan replied:
Good idea. I've created the following types:
- `DirectionalAppTaggingSecret`
- `Tag`
- `SiloedTag`

I didn't create the `TaggingSharedSecret` type because it's never needed out of the initial computation. But I have a `computeSharedTaggingSecret` in the `directional_app_tagging_secret.ts` file that is not exposed and is only used inside the `DirectionalAppTaggingSecret` constructor. Given that the shared secret doesn't need to travel around introducing the type for it felt unnecessary.

The only sad thing is that the `DirectionalAppTaggingSecret` needs to reside in stdlib as it's part of the IndexedTaggingSecret while the rest is in PXE.

benesjan commented on yarn-project/pxe/src/contract_function_simulator/tagging_utils.ts:
```diff
@@ -1,32 +1,74 @@
 // Half the size of the window we slide over the tagging secret indexes.
 export const WINDOW_HALF_SIZE = 10;
 
+// TODO(benesjan): Make this return tags instead - this will moves some complexity from syncTaggedLogs
```
Will do this in a follow-up PR. It will make the code a bit nicer.

nventuro commented on yarn-project/pxe/src/storage/tagging_data_provider/tagging_data_provider.ts:
```diff
@@ -8,86 +8,54 @@ export class TaggingDataProvider {
+  // Stores the last index used for each directional app tagging secret, taking into account whether we are requesting
+  // the index as a sender or as a recipient because the sender and recipient can be in the same PXE.
+  #taggingSecretIndexesAsSenders: AztecAsyncMap<string, number>;
+  #taggingSecretIndexesAsRecipients: AztecAsyncMap<string, number>;
```
Took me a while to figure what we were actually storing, should this be called 'lastUsedIndex` instead of just index?

benesjan replied:
It was actually a lie as it stores the next secrets (to be included in a log when sender or to look for when syncing logs as recipient).

Did a larger renaming here and renamed `#taggingSecretIndexesAsSenders` as `#nextIndexesAsSenders`. The functionality here resides in `TaggingDataProvider` so I think the "taggingSecret" in the name is completely unnecessary (while having "next" there is genuinely useful).

nventuro commented on yarn-project/pxe/src/storage/tagging_data_provider/tagging_data_provider.ts:
```diff
@@ -8,86 +8,54 @@ export class TaggingDataProvider {
-      indexedSecrets.map(indexedSecret =>
-        storageMap.set(`${indexedSecret.appTaggingSecret.toString()}_${inDirectionOf.toString()}`, indexedSecret.index),
+      indexedTaggingSecrets.map(({ directionalAppTaggingSecret, index }) =>
+        this.#taggingSecretIndexesAsSenders.set(directionalAppTaggingSecret.toString(), index),
```
It sounds we like we should error if two entries in the array contain the same secret with different indexes, no? Which gets applied last would be undefined as it'd depend on the internal timings of the async functions. So maybe we should check that this has not happened?

benesjan replied:
Good idea. I think we can make it even stricter and error if the input contains 2 equal secrets as that should never happen in PXE. Implemented in https://github.com/AztecProtocol/aztec-packages/pull/17451/commits/bf116f2c0b4b6c05d0ffe96f37eeb6097d5e1447

benesjan commented on yarn-project/stdlib/src/logs/directional_app_tagging_secret.ts:
```diff
@@ -0,0 +1,61 @@
+   * Derives shared tagging secret and from that, the app address and recipient derives the directional app tagging
+   * secret.
+   *
+   * @param localAddress - The complete address of entity 1 in the shared tagging secret derivation scheme
+   * @param localIvsk - The incoming viewing secret key of entity 1
```
Compared to the original code I prefixed these 2 with "local" to communicate that they correspond to 1 entity. Then I think it's clearer that external corresponds to the entity 2.

benesjan commented on yarn-project/stdlib/src/logs/directional_app_tagging_secret.ts:
```diff
@@ -0,0 +1,68 @@
+  externalAddress: AztecAddress,
+): Promise<Point> {
+  const knownPreaddress = await computePreaddress(await localAddress.publicKeys.hash(), localAddress.partialAddress);
+  // TODO: #8970 - Computation of address point from x coordinate might fail
```
Copied this TODO here with the original code. Josie tackled that issue in [this PR](https://github.com/AztecProtocol/aztec-packages/pull/17134) and I think we can afford to remove it from here as it seems fine that we throw in the `toAddressPoint` function on the next line if the address is invalid here.

Agree?

benesjan commented:
@nventuro managed to not rebase this with graphite so looking for changes since your last review should work.

benesjan committed:
refactor: tagging related cleanup

I started working on improving unconstrained tagging as proposed in [this forum post](https://forum.aztec.network/t/on-note-discovery-and-index-coordination/7165) and the initial step in implementing that is the following:

1. store the tagging indexes requested during execution in a cache (just like we do with notes),
2. when the execution is finished take the cache and along with a tx hash dump it into db.

Given that the cache will reside in `PrivateExecutionOracle` I now need to orchestrate the sync there.

When trying to do that (to be done in a [PR up the stack](https://github.com/AztecProtocol/aztec-packages/pull/17445)) I realized that the current tagging-related API was cumbersome and this made me do the following in this PR:
1. Have the API accept directionalAppTaggingSecret on the input instead of appTaggingSecret and recipient as now I can just compute it in the `PrivateExecutionOracle` and request info based on that,
2. I moved the tagging functionality from stdlib to PXE as there was no reason to have it in stdlib,
3. I simplified a bunch of random things.

This PR is a bit random so if something is not clear LMK.
